bazel_repositories_priv

```diff
@@ -0,0 +1,292 @@
+    namespace without Internet connectivity.
+    """
+    log.debug(f"sub target: {target}")
+    os.chdir(root_dir)
+
+    nsfd = os.open(os.path.join("/var/run/netns", netns_name), os.O_RDONLY)
+    libc.setns(nsfd, CLONE_NEWNET)
+
+    as_user()
+
+    args = ["bazel", "fetch", "--repository_cache=../BAZEL_CACHE"]
+    if bazel_args:
+        args += bazel_args.split()
+    args += [target]
+
+    log.debug(f"args: {args}")
+
+    p = subprocess.Popen(args, stdout=subprocess.PIPE,
+                         stderr=subprocess.STDOUT,
+                         universal_newlines=True)
+    stdout = ""
+    for line in iter(p.stdout.readline, ""):
+        line_s = line.strip()
+        log.debug(f"bazel fetch: {line_s}")
+        stdout += line
+    p.stdout.close()
+    returncode = p.wait()
+
+    urls = re.findall(r"https?://(?:[a-zA-Z]|[0-9]|[-_@.&/+])+", stdout)
+    urls = list(dict.fromkeys(urls))
+
+    conn.send((returncode, urls,))
+    conn.close()
+
+
+def chown(pathname: str):
+    """Change the owner of the given file from root to the unprivileged user.
+    """
+    user_passwd = pwd.getpwnam(user)
+    log.debug(f"chown {pathname}")
+    os.chown(pathname, user_passwd.pw_uid, user_passwd.pw_gid)
+    for root, dirs, files in os.walk(pathname):
+        for d in dirs:
+            filename = os.path.join(root, d)
+            log.debug(f"chown {filename}")
+            os.chown(filename, user_passwd.pw_uid, user_passwd.pw_gid)
+        for f in files:
+            filename = os.path.join(root, f)
+            log.debug(f"chown {filename}")
+            os.chown(filename, user_passwd.pw_uid, user_passwd.pw_gid)
+
+
+def download(url: str, dst: str):
+    log.info(f"Downloading {url}")
+    urllib_request.urlretrieve(url, dst)
+    log.info("Download finished")
+
+
+def process_url(url: str, url_path: pathlib.Path, filename: str) -> str:
+    """Download and store the archive from the given URL.
+    """
+    # If two first chars are not alpha, it means that the tarball name is a
+    # version or a commit hash. In that case, append the project name for
+    # readability.
+    if not (filename[0].isalpha() and filename[1].isalpha()):
+        prefix = None
+        # URLs from mirror.bazel.build have path like i.e.:
+        # mirror.bazel.build/github.com/account/project
+        # so the project name is the 3rd element of the path (which does not
+        # include the domain).
+        if "mirror.bazel.build" in url:
+            prefix = url_path.parts[3]
+        # And we assume that the other tarballs are downloaded from git hosting
+        # services, where path usually looks like i.e.:
+        # github.com/account/project
+        # gitlab.com/account/project
+        # so the project name is the 2nd element of the path (which does not
+        # include the domain).
+        else:
+            prefix = url_path.parts[2]
+        filename = prefix + "-" + filename
```

Comment:
No f-string here?  I'm torn on whether `f"{prefix}-{filename}"` is easier or harder to read.  Just thinking out loud, I guess.

Reply:
This code will be removed too.